static/app/utils/profiling/profile/chromeTraceProfile.tsx

```diff
@@ -395,22 +395,139 @@ function collectEventsByProfile(input: ChromeTrace.ArrayFormat): {
+
+    if (typeof profileNode.parent === 'number') {
+      const parent = frames.get(profileNode.parent);
+
+      if (parent === undefined) {
+        throw new Error('Missing frame parent in profile');
+      }
+    }
+
+    if (!profileNode.children) {
+      continue;
+    }
+
+    for (let j = 0; j < profileNode.children.length; j++) {
+      const child = frames.get(profileNode.children[j]);
+
+      if (child === undefined) {
+        throw new Error('Missing frame child in profile');
+      }
+
+      child.parent = profileNode;
+    }
+  }
+
+  return frames;
+}
+
+// Cpu profiles can often contain a lot of sequential samples that point to the same stack.
+// It's wasteful to process these one by one, we can instead collapse them and just update the time delta.
+// We should consider a similar approach for the backend sample storage. I expect we will remove
+// this code from the frontend once we have backend support and a unified format for these.
+// Effectively, samples like [1,1,2,1] and timedeltas [1,2,1,1] to sample [1,2,1] and timedeltas [3,1,1]
+export function collapseSamples(profile: ChromeTrace.CpuProfile): {
+  sampleTimes: number[];
+  samples: number[];
+} {
+  const samples: number[] = [];
+  const sampleTimes: number[] = [];
+
+  // If we have no samples, then we can't collapse anything
+  if (!profile.samples || !profile.samples.length) {
+    throw new Error('Profile is missing samples');
+  }
+
+  // If we have no time deltas then the format may be corrupt
+  if (!profile.timeDeltas || !profile.timeDeltas.length) {
+    throw new Error('Profile is missing timeDeltas');
+  }
+
+  // If timedeltas does not match samples, then the format may be corrupt
+  if (profile.timeDeltas.length !== profile.samples.length) {
+    throw new Error("Profile's samples and timeDeltas don't match");
+  }
+
+  if (profile.samples.length === 1 && profile.timeDeltas.length === 1) {
+    return {samples: [profile.samples[0]], sampleTimes: [profile.timeDeltas[0]]};
+  }
+
+  // First delta is relative to profile start
+  // https://github.com/v8/v8/blob/44bd8fd7/src/inspector/js_protocol.json#L1485
+  let elapsed: number = profile.timeDeltas[0];
+
+  // This is quite significantly changed from speedscope's implementation.
+  // We iterate over all samples and check if we can collapse them or not.
+  // A sample should be collapsed when there are more that 2 consecutive samples
+  // that are pointing to the same stack.
+  for (let i = 0; i < profile.samples.length; i++) {
+    const nodeId = profile.samples[i];
+
+    // Initialize the delta to 0, so we can accumulate the deltas of any collapsed samples
+    let delta = 0;
+    // Start at i
+    let j = i;
+    // While we are not at the end and next sample is the same as current
+    while (j < profile.samples.length && profile.samples[j + 1] === nodeId) {
+      // Update the delta and advance j
+      delta = Math.max(delta + profile.timeDeltas[j + 1], delta);
```

Comment:
Can we leave a comment explaining the negative time deltas and why this isn't just `delta += profile.timeDeltas[j + 1]`.